mc/Mongo-SDAM.package/MongoPool.class/instance/mongoDo..st

```diff
@@ -0,0 +1,14 @@
+accessing
+mongoDo: aBlockClosure
+	"Evaluate the BlockClosure with a Mongo instance. See #next.
+	
+	For error handling, the specification states that the client MUST clear its connection pool for the server: if one socket is bad, it is likely that all are.
+
+	Source: https://github.com/mongodb/specifications/blob/573b1f58a129056d651781ad99317b6c656e050e/source/server-discovery-and-monitoring/server-discovery-and-monitoring-summary.rst#id9
+	"
+
+	[ | mongo |
+		mongo := self next.
+		^ [ aBlockClosure value: mongo ] ensure: [ self return: mongo ] ]
```

Comment:
The unconditional "self return: mongo" is dangerous. I got it to break int he following way (with Voyage):

1. Read once (to create a connection)
2. Use iptables -j DROP to simulate full packet loss
3. Read and let it run into the configured timeout
4. Remove the netfilter ("iptables") blocking..
5. Read again.
6. Read again.
7. ...

What happens is that:

* In 3.) we wrote to the socket but never saw the Mongo Query Reply. TCP can keep connection open a long time and has good error recovery. Once 4.) happens the request will be sent to the server and a reply will be queued. But at this point no MongoCursor>>#execute is running.

* In 5.) We send a query and read the _previous_ reply. An error is signalled for this case. The Mongo is then returned to the pool.
* In 6.) We send a query and read the _previous_ reply...
* And this continues forever...


WDUT about:

1. Maybe use ifCurtailed to close the mongo and return to the pool inside the block.
2. The >>#on:do will be called before >>#ensure: This means we empty the pool for connections that might not be broken... and then return the broken one to the pool.
3. I would probably not break connections in this way? E.g. we might see a timeout because a query is very complicated?

Reply:
I see, you detected a bug. The wrong reply is found in MongoCursor>>execute like:
```
	reply responseTo = requestId ifFalse: [
		^self error: 'Unexpected responseTo in response' ].
```
This error may deserve to have its own class and then ease catching it. The pool contains Mongo instances that are open and in good status, ready to be used next time a mongoDo: requires it. I will have to answer more deeply with some code tests in next days.

Reply:
That's a check I added and I agree I should have made it class based. With errors like this I prefer a pessimistic approach to discard the Mongo object and create a new one.

Reply:
The fix I propose in commit 8d2260b uses a new error class for the 'Unexpected responseTo in response' and `mongoDo:` catches it to act like in case of NetworkError: close the connection and pass the error for next handler.
The catch acts only in those two errors because they are the ones that inhibit the Mongo connection to be reused in a next operation on the pool (because of it's in an inconsistent state). Hmm, now that I think on it, the connection close should be non-blocking to speedup this scenario (no?). 
The idea of the pool of Mongo instances is that is only has Mongo instances that are open and so ready to be used in next operation saving the time to open a new one.

The MongoPool is still missing a limit on the max number of connections it opens. It's currently unlimited, user asks for mongoDo: from a process and it will try to open a Mongo immediately. IIRC according to SDAM reference, the process should be pu tto wait until a Mongo instance is avalable when the count exeeds a limit.

Reply:
The code seems unsafe. Any unhandled exception will leak the Mongo object now and then we go out of external semaphores?